README.md
# SigmaCapital
 A revolutionary new approach to stock investing through data driven recommendations. SigmaCapital seeks to empower you to make informed financial decisions using the power of data analytics.

# Table of Contents
* [Introduction](https://github.com/AbChatt/SigmaCapital#introduction)
* [Technologies](https://github.com/AbChatt/SigmaCapital#technologies)
* [Requirements](https://github.com/AbChatt/SigmaCapital#requirements)
* [Status](https://github.com/AbChatt/SigmaCapital#status)
* [Acknowledgements](https://github.com/AbChatt/SigmaCapital#acknowledgements)

## Introduction
Investing into the stock market has always beeen a daunting task. For first time buyers, there are many pitfalls to avoid. Which stocks should I invest in? What current market trends do I need to be aware of? When should I sell my shares? These are common concerns that can arise when dipping your toes into the stock market. SigmaCapital was conceived as a response to these challenges and seeks to simplify the convoluted world of finance and stock market investing. We use advanced data analytics in order to generate real time predictions about the near future value of a stock, giving you the insights you need in order to make informed decisions on the stock market and maximise your returns. 

We've made a video about it. Check it out [here](https://youtu.be/NdpAy7NElLA)!

## Technologies
The app runs on a Python backend, which samples 5 year historical data from the Alpha Vantage API. This data is fed into a set of models that generate predictions, which are then combined with information from other economic indicators in order to produce a final recommendation of "Buy", "Sell" or "Hold". This recommendation is then streamed to a Initial State Front End.

## Requirements
Python 3.7 is the main requirement. As such, you will also need to pip install the latest versions of alpha-vantage, ISStreamer and pmdarima (which fetches statsmodels, which is also required).

## Status
As of 7th June 2020, the project status is still active. For future improvements, we want to add more sophisticated models and indicators to our repertoire so that we can further optimise the accuracy of predictions. We also want to add support for indices such as the S&P500 and FTSE100 and generate sector specific predictions. We are also exploring plans to integrate our product into banking accounts, so we can generate personalised predictions for our customers. Our vision is to create a comprehensive platform that encompasses stocks, bonds, ETFs, foreign exchange and cryptocurrencies. This is just the beginning.

## Acknowledgements
We used the [Alpha Vantage API](https://github.com/RomelTorres/alpha_vantage) in order to get the latest, up-to-date stock data. Furthermore, parts of the project were adapted from [here](https://towardsdatascience.com/performing-a-time-series-analysis-on-the-aapl-stock-index-3655da9612ff).

changelog.txt
Alpha releases
Initial build
Changed plot style to dark_background
Fixed closing prices graph - now working
Fixed mean closing price vs month - now working
Fixed train and test sets
Renamed CSV

V2 (Lots of changes)

Had some mistakes in the training and test sets - now completely working
Created a graph to show training vs. test set, so TODO has been removed
Added title to moving average graph
Added graph labels to moving average and several others
Added prediction code back, added prediction line to graph
Fixed axis for prediction model, so now it fits in with the overall graph

Beta releases

V 0.7

Added Root Mean Squared code (blueprint, will be modified later with more complex logic)
Added BUY/SELL rating functionality 
General code cleanup/commenting
Added next several TODOs

V 0.8

Added conditional stationary checks
Commented-out an incessant number of graphs and prints
Cleaned some code
Reorganized some code

V 0.9

Changed prediction line color on graph
Added ticker data to graphs
Encapsulated model generation into one function
Cleaned up code

V 1.0

Created modelSelection to deal with multiple hypothetical models
Created main.py
Added model slope to function output

V 1.0.1

Created a specific rating function for the close price model
Attempt at implementation of Holt’s exponential damping algorithm - HAS BUGS

V 1.1

Implemented Holt’s exponential damping algorithm
Added Holt’s model to the model evaluation function

V 1.2

Model.py
Implemented model selection logic
Added ticker detail to rating

Main.py
Implemented Top 5 / Bottom 5 logic

V 1.3

Model.py

Officially migrated to Alpha Vantage Stocks API
Expanded historical data to 5 years
Major performance optimizations
Changed training and test sets
Added todos

Main.py

Added functionality to the Top 5/ Bottom 5 Logic

V 1.4

Model.py
Added asynchronous data retrieval
Added Golden Cross / Death Cross metric calculation

Main.py
Updated to support calling API asynchronously
Encapsulated API key

V 1.5

Model.py
Added more complex decision logic for Golden/Death cross
Added a third decision option - HOLD. 

Main.py

Added Dashboard support

main.py
from model import *
import asyncio
from ISStreamer.Streamer import Streamer

# print(modelSelection('AAPL_Jun_2019_2020.csv', "AAPL"))
# print(modelSelection('AAPL-2.csv', "AAPL"))

# list of companies
# tickers = ['AAPL', 'DIS', 'XOM']
# tickers = ['AAPL', 'DIS', 'XOM']
tickers = input("Please enter the tickers of the stocks you want, seperated by a space: ").split()
print(tickers)
def highlights(ratings, rating):
	buy_and_sell = sorted(ratings, key=lambda x: x[0])
	buy = sorted(list(filter(lambda x: x[0] == "BUY", buy_and_sell)), key=lambda x: x[1])
	sell = sorted(list(filter(lambda x: x[0] == "SELL", buy_and_sell)), key=lambda x: x[1])
	our_highlights = (list(reversed(buy[-5:])), list(sell[0:5]))
	if rating == "BUY":
		return [x[3] for x in our_highlights[0]]
	else:
		return [x[3] for x in our_highlights[1]]

# get API key
print("Please enter your API key: ")
api_key = input()

# asynchronously get input
loop = asyncio.get_event_loop()
tasks = [get_data(ticker, api_key) for ticker in tickers]
group1 = asyncio.gather(*tasks)
results = loop.run_until_complete(group1)


# Configure Streaming Bucket

ACCESS_KEY = "ist_ZmIGaAlTd9_qv0JcanHg3XzKA_Y7Y284"
BUCKET_KEY = "3DQY5TQB3CUB"
BUCKET_NAME = "Hack The Northeast"
streamer = Streamer(bucket_name=BUCKET_NAME, bucket_key=BUCKET_KEY, access_key=ACCESS_KEY)

print("And the results are: ")
for i in range(len(results)):
	# print(modelSelection(results[i][0], tickers[i]))
	now = modelSelection(results[i], tickers[i])
	streamer.log(tickers[i], now[0])
	print(now)
	
streamer.flush()


# print(highlights([("SELL", -1, 1, "GOOGL"), ("SELL", -0.5, 1, "RDSA"), ("BUY", 100, 1, "APPL"), ("BUY", 80, 1, "BP"), ("SELL", -0.3, 1, "XXON"), ("BUY", 70, 1, "MSFT"), ("BUY", 60, 1, "BRK.A"), ("BUY", 50, 1, "JPM"), ("SELL", -0.4, 1, "CMCSA"), ("SELL", -0.2, 1, "PM")], "SELL"))

astro.py
import urllib3
import json
import os
from datetime import datetime


#--- WEATHER PROGRAMMING ------ #

CITY = "MAIDENHEAD"
GPS_COORDS = "51.526301,-0.731486"
DARKSKY_API_KEY = "9b88e268bff31516691599141aefa11c"


def get_current_conditions():
	api_conditions_url = "https://api.darksky.net/forecast/" + DARKSKY_API_KEY + "/" + GPS_COORDS + "?units=auto"
	try:
		http = urllib3.PoolManager()
		r = http.request('GET', api_conditions_url)
	except:
		return []
	json_currently = r.data
	return json.loads(json_currently)
	
def moon_icon(moon_phase):
	if moon_phase < .06:
		return "Tonight is a new moon"
	if moon_phase < .125 and moon_phase >= 0.06:
		return "Tonight, the moon is a waxing crescent. "
	if moon_phase < .25 and moon_phase >= 0.125:
		return "Tonight, the moon is in its first quarter. "
	if moon_phase < .48 and moon_phase >= 0.25:
		return "Tonight, the moon is a waxing gibbbous. "
	if moon_phase < .52 and moon_phase >= 0.48:
		return "Tonight, is a full moon."
	if moon_phase < .625 and moon_phase >= 0.52:
		return "Tonight, the moon is a waning gibbous. "
	if moon_phase < .75 and moon_phase >= 0.625:
		return "Tonight, the moon is in its last quarter. "
	if moon_phase < 1 and moon_phase >= 0.75:
		return "Tonight, the moon is a waning crescent. "

def isFloat(string):
	try:
		float(string)
		return True
	except ValueError:
		return False
		
curr_conditions = get_current_conditions()
hours = [i for i in [j['time'] for j in curr_conditions['hourly']['data']]]

sunset = curr_conditions['daily']['data'][0]['sunsetTime']

sunset_min = str(datetime.fromtimestamp(sunset).minute) if len(str(datetime.fromtimestamp(sunset).minute)) > 1 else "0" + str(datetime.fromtimestamp(sunset).minute)

sunrise = curr_conditions['daily']['data'][1]['sunriseTime']

sunrise_min = str(datetime.fromtimestamp(sunrise).minute) if len(str(datetime.fromtimestamp(sunrise).minute)) > 1 else "0" + str(datetime.fromtimestamp(sunrise).minute)

today = datetime.fromtimestamp(hours[0]).day
moon_phase = moon_icon(curr_conditions['daily']['data'][0]['moonPhase'])

relevant_hours = [i for i in hours if sunset < i and i <= sunrise]

cloud_list = [(datetime.fromtimestamp(hour['time']).hour,str(round(hour['cloudCover']*100))+"%") for hour in [i for i in curr_conditions['hourly']['data']] if hour['time'] in relevant_hours and hour['cloudCover'] <= 0.1]


output = ""
if len(cloud_list) == 0:
	output = "No, the skies are far too cloudy today. Try again tomorrow."
else:
	output += "Yes, you can observe tonight. "
	for tupl in cloud_list:
		if tupl[0] > 12:
			output += f"Cloud cover is {tupl[1]} at {tupl[0] % 12} PM. "
		elif tupl[0] == 0:
			output += f"Cloud cover is {tupl[1]} at 12 AM. "
		else:
			output += f"Cloud cover is {tupl[1]} at {tupl[0]} AM. "
	output += f"Sunset tonight is at {datetime.fromtimestamp(sunset).hour}:{sunset_min} PM. "
	output += moon_phase + f"Moon prescence is at {round(curr_conditions['daily']['data'][0]['moonPhase']*100)}%. "	
	output += f"Sunrise tomorrow is at {datetime.fromtimestamp(sunrise).hour}:{sunrise_min} AM."

#---- CLIPBOARD ACTION ------ #



model.py
import os
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math
import asyncio

from pylab import rcParams
from datetime import datetime
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.arima_model import ARIMA
from statsmodels.tsa.holtwinters import ExponentialSmoothing

# needed to pip install pmdarima - which also fetched statsmodels
from pmdarima.arima import auto_arima
from sklearn.metrics import mean_squared_error, mean_absolute_error

# import API
from alpha_vantage.timeseries import TimeSeries
from alpha_vantage.async_support.timeseries import TimeSeries




# Define the buy/sell rating as appropriate
def rating(prediction):
    number_of_days = len(prediction) # How many days are in our test set
    initial_price = prediction['Prediction'].iloc[0] # Variable names for these
    final_price = prediction['Prediction'].iloc[-1] # are self explanatory
    model_slope = (prediction['Prediction'].iloc[-1] - initial_price)/number_of_days # Basic linear regression

    if model_slope > 0.01:
        return ["BUY", model_slope]
    elif round(model_slope) == 0 and model_slope > 0:
        return["HOLD", model_slope]
    else:
        return ["SELL", model_slope]

# Determine if timeseries is stationary
def test_stationarity(timeseries):
    # Determining rolling statistics
    roll_mean = timeseries.rolling(12).mean()
    roll_std = timeseries.rolling(12).std()

    # Plot rolling statistics
    # plt.plot(timeseries, color='blue', label='Original')
    # plt.plot(roll_mean, color='red', label='Rolling Mean')
    # plt.plot(roll_std, color='white', label='Rolling Std')
    # plt.legend(loc='best')
    # plt.title('Rolling Mean and Standard Deviation')
    # plt.show()  # using block = False stops graph from showing, but block = True seems to be the same as not including it at all

    # running Augmented Dickey Fuller Test
    # print("Results of Dickey Fuller test")
    adft = adfuller(timeseries, autolag='AIC')

    # adfuller gives output without labels - need to manually write what values mean using for loop
    output = pd.Series(adft[0:4], index=['Test Statistics', 'p-value', 'No. of lags used', 'Number of observations used'])
    for key, values in adft[4].items():
        output['critical value (%s)'%key] = values

    test_stats = output[0] # Test statistics
    critical_values = output[4] # Critical values

    if test_stats > critical_values: # If the test statistics are more than the critical values, the null hypothesis passes and it is not stationary
        return False
    else:
        return True # The null hypothesis IS stationary

def generateCloseModel(data_frame, ticker):

    # split dataset into testing and training sets - This wasn't working, now has been fixed
    limit = int(0.9*data_frame.shape[0])
    test = data_frame[limit:] # last 20% of values
    train = data_frame[:limit] # first 80% of values
    # Plot into graph
    # plt.plot(train.groupby(['date']) ['4. close'].mean(), color='blue', label='Train')
    # plt.plot(test.groupby(['date']) ['4. close'].mean(), color='red', label='Test')
    # plt.legend(loc='best')
    # plt.title('Training and Test Data')
    # plt.show()

    is_stationary = test_stationarity(train['4. close'])

    #-- SUMMARY OF LINES 118 - 194 -------------
    # If the stock prices are not stationary, we remove the trends and seasonality

    if not is_stationary:
    # taking ln of training and testing closing values
        train_log = np.log(train['4. close'])
        test_log = np.log(test['4. close'])

        # plotting moving average
        moving_avg = train_log.rolling(24).mean()
        # plt.plot(train_log, label='Original')
        # plt.plot(moving_avg, color='red', label='Moving Average')
        # plt.title('Moving Average')
        # plt.show()

        # removing trend to make time series stationary
        train_log_moving_avg_diff = train_log - moving_avg
        train_log_moving_avg_diff.dropna(inplace=True)  # took average of first 24 values, so rolling mean is not defined for first 23 - drop values

    # test_stationarity(train_log_moving_avg_diff)

        # stabilise mean of time series - required for stationary time series
        train_log_diff = train_log - train_log.shift(1)
        test_stationarity(train_log_diff.dropna())


    # fit auto ARIMA - fit model on the univariate series

        model = auto_arima(train_log, trace=True, error_action='ignore', suppress_warnings=True)
        model.fit(train_log)

        # predict values on validation set
        forecast = model.predict(n_periods=len(test))  # n_periods creates errors
        forecast = pd.DataFrame(forecast, index=test_log.index, columns=['Prediction'])
        forecast = np.exp(forecast) # Adjust for non-log values


        # Graph forecasts vs actual
        # plt.plot(train.groupby(['date']) ['4. close'].mean(), label='Train')
        # plt.plot(test.groupby(['date']) ['4. close'].mean(), label='Test')
        # plt.plot(forecast, color='red', label='Prediction')
        # plt.title(f'{ticker} Stock Price Prediction')
        # plt.xlabel('Time')
        # plt.ylabel('Actual Stock Price')
        # plt.legend(loc='upper left', fontsize=8)
        # plt.show()

        # check performance of model using RMSE as metric

        rms = np.sqrt(mean_squared_error(test_log,np.log(forecast)))

    else:
        model = auto_arima(train['4. close'], trace=True, error_action='ignore', suppress_warnings=True)
        model.fit(train['4. close'])

        # predict values on validation set
        forecast = model.predict(n_periods=len(test))  # n_periods creates errors
        forecast = pd.DataFrame(forecast, index=test.index, columns=['Prediction'])

        # Graph forecasts vs actual
        # plt.plot(train.groupby(['date']) ['4. close'].mean(), label='Train')
        # plt.plot(test.groupby(['date']) ['4. close'].mean(), label='Test')
        # plt.plot(forecast, color='red', label='Prediction')
        # plt.title(f'{ticker} Stock Price Prediction')
        # plt.xlabel('Time')
        # plt.ylabel('Actual Stock Price')
        # plt.legend(loc='upper left', fontsize=8)
        # plt.show()

        # check performance of model using RMSE as metric

        rms = np.sqrt(mean_squared_error(test,forecast))

    result = rating(forecast)
    result.append(rms)
    result.append(ticker)
    return tuple(result)


    # print(f" You should {rating(forecast)} this stock")

def HoltExp(data_frame, ticker):

    # split dataset into testing and training sets - This wasn't working, now has been fixed
    limit = int(0.9*data_frame.shape[0])
    test = data_frame[limit:] # last 20% of values
    train = data_frame[:limit] # first 80% of values
    # Plot into graph
    # plt.plot(train.groupby(['date']) ['4. close'].mean(), color='blue', label='Train')
    # plt.plot(test.groupby(['date']) ['4. close'].mean(), color='red', label='Test')
    # plt.legend(loc='best')
    # plt.title('Training and Test Data')
    # plt.show()

    is_stationary = test_stationarity(train['4. close'])

    #-- SUMMARY OF LINES 118 - 194 -------------
    # If the stock prices are not stationary, we remove the trends and seasonality

    if not is_stationary:
        # taking ln of training and testing closing values
        train_log = np.log(train['4. close'])
        test_log = np.log(test['4. close'])

        # plotting moving average
        moving_avg = train_log.rolling(24).mean()
        # plt.plot(train_log, label='Original')
        # plt.plot(moving_avg, color='red', label='Moving Average')
        # plt.title('Moving Average')
        # plt.show()

        # removing trend to make time series stationary
        train_log_moving_avg_diff = train_log - moving_avg
        train_log_moving_avg_diff.dropna(inplace=True)  # took average of first 24 values, so rolling mean is not defined for first 23 - drop values

        # test_stationarity(train_log_moving_avg_diff)

        # stabilise mean of time series - required for stationary time series
        train_log_diff = train_log - train_log.shift(1)
        test_stationarity(train_log_diff.dropna())


        # fit Exponential Smoothing - fit model on the univariate series

        model = ExponentialSmoothing(np.array(train_log), trend='mul', seasonal=None, damped=True)
        fit = model.fit()

        # predict values on validation set
        forecast = fit.forecast(np.asarray(len(test)))  # n_periods creates errors
        forecast = pd.DataFrame(forecast, index=test_log.index, columns=['Prediction'])
        forecast = np.exp(forecast) # Adjust for non-log values

        # Graph forecasts vs actual
        # plt.plot(train.groupby(['date']) ['4. close'].mean(), label='Train')
        # plt.plot(test.groupby(['date']) ['4. close'].mean(), label='Test')
        # plt.plot(forecast, color='red', label='Prediction')
        # plt.title(f'{ticker} Stock Price Prediction')
        # plt.xlabel('Time')
        # plt.ylabel('Actual Stock Price')
        # plt.legend(loc='upper left', fontsize=8)
        # plt.show()

        # check performance of model using RMSE as metric

        rms = np.sqrt(mean_squared_error(test_log,np.log(forecast)))

    else:
        model = ExponentialSmoothing(np.array(train['4. close']), trend='mul', seasonal=None, damped=True)
        fit = model.fit()

        # predict values on validation set
        forecast = fit.forecast(len(test))  # n_periods creates errors
        forecast = pd.DataFrame(forecast, index=test.index, columns=['Prediction'])

        # Graph forecasts vs actual
        # plt.plot(train.groupby(['date']) ['4. close'].mean(), label='Train')
        # plt.plot(test.groupby(['date']) ['4. close'].mean(), label='Test')
        # plt.plot(forecast, color='red', label='Prediction')
        # plt.title(f'{ticker} Stock Price Prediction')
        # plt.xlabel('Time')
        # plt.ylabel('Actual Stock Price')
        # plt.legend(loc='upper left', fontsize=8)
        # plt.show()

        # check performance of model using RMSE as metric

        rms = np.sqrt(mean_squared_error(test,forecast))

    result = rating(forecast)
    result.append(rms)
    result.append(ticker)
    return tuple(result)


def goldenDeathCross(data_frame, ticker):

    # split dataset into testing and training sets - This wasn't working, now has been fixed
    limit = int(0.9*data_frame.shape[0])
    test = data_frame[limit:] # last 20% of values
    train = data_frame[:limit] # first 80% of values
    # Plot into graph
    # plt.plot(train.groupby(['date']) ['4. close'].mean(), color='blue', label='Train')
    # plt.plot(test.groupby(['date']) ['4. close'].mean(), color='red', label='Test')
    # plt.legend(loc='best')
    # plt.title('Training and Test Data')
    # plt.savefig(f"{ticker}_TrainTest.png")
    #plt.show()

    is_stationary = test_stationarity(train['4. close'])

    #-- SUMMARY OF LINES 118 - 194 -------------
    # If the stock prices are not stationary, we remove the trends and seasonality

    # taking ln of training and testing closing values
    train_log = np.log(train['4. close'])
    test_log = np.log(test['4. close'])

    # plotting moving average
    moving_avg_200 = train_log.rolling(200).mean()
    moving_avg_50 = train_log.rolling(50).mean()
    # plt.plot(train_log, label='Original')
    # plt.plot(moving_avg_200, color='red', label='200 Day Moving Average')
    # plt.plot(moving_avg_50, color='green', label='50 Day Moving Average')
    # plt.title('Moving Average')
    # plt.legend(loc='upper left', fontsize=8)
    # plt.show()


    # removing trend to make time series stationary
    train_log_moving_avg_diff = train_log - moving_avg_200
    train_log_moving_avg_diff.dropna(inplace=True)  # took average of first 24 values, so rolling mean is not defined for first 23 - drop values
    
    # stabilise mean of time series - required for stationary time series
    train_log_diff = train_log - train_log.shift(1)
    test_stationarity(train_log_diff.dropna())

    #print(test)

    # checking values of moving average curves near split of train and test sets
    #moving_avg_200_value = (np.log((data_frame[limit - 198:limit + 2])['4. close'])).rolling(200).mean()
    #moving_avg_50_value = (np.log((data_frame[limit - 48:limit + 2])['4. close'])).rolling(50).mean()
    
    moving_avg_200.dropna(inplace=True)

    moving_avg_50.dropna(inplace=True)
    #moving_avg_50.drop(moving_avg_50[moving_avg_50['date'] < moving_avg_200['date']].index, axis=1, inplace=True)
    #print(moving_avg_50.index)

    #print(data_frame[:limit + 1].index[-1])
    #boundary = (data_frame[:limit + 1].index[-1]).strftime("%Y-%m-%d")
    #print(boundary)
    #print(type(boundary))

    #print(moving_avg_200.index)
    #print(moving_avg_50.index)
    #print(moving_avg_200[0])
    #moving_avg_200_filtered = moving_avg_200[moving_avg_200.index >= boundary]
    #print(moving_avg_200_filtered)
    #print(moving_avg_200.index)
    #print(moving_avg_50_value)

    if (moving_avg_50[-1] > moving_avg_200[-1]):
        return "BUY"
    else:
        return "SELL"

def modelSelection(data_frame, ticker):
    # inserts an entry into list of warning filters
    warnings.filterwarnings('ignore')
    
    # style of graph being plotted - font style and related preferences (can be changed)
    plt.style.use('dark_background')
    
    # changing default figsize parameter
    rcParams['figure.figsize'] = 10, 6

    #print("Please enter your API key: ")
    #api_key = input()
    
    #ts = TimeSeries(key=api_key, output_format='pandas')
    
    
    #data_frame, meta_data = ts.get_daily(symbol=ticker, outputsize='full')
    # pd.read_csv(data)    # modified from example as path was not being picked up
    
    # gets Date column data
    # con = data_frame['date']
    # converts date column to a Python DateTime object
    data_frame['date'] = pd.to_datetime(data_frame.index)
    
    # sets data_frame index to Date column
    data_frame.set_index('date', inplace=True)
    
    # get Year, Month and Day columns
    data_frame['year'] = data_frame.index.year
    data_frame['month'] = data_frame.index.month
    data_frame['day'] = data_frame.index.day
    data_frame = data_frame.sort_index(axis=0 ,ascending=True) # Reverse DataFrame
    indexNames = data_frame[data_frame['year'] < datetime.now().year - 5].index
    data_frame.drop(indexNames , inplace=True)
    # Display a random sample of 5 rows - year, month and day columsn are added
    #print(data_frame.sample(5, random_state=0))
    
    # groups data by date and close - x axis is date, y axis is mean closing price
    temp = data_frame.groupby(['date']) ['4. close'].mean()
    
    # plots line graph of mean closing price vs date - WORKING NOW
    # temp.plot(figsize=(15, 5), kind='line', title=f'{ticker} Closing Prices', fontsize=14)
    # plt.savefig(f"{ticker}.png")
    # plt.show()
    
    # plots bar graph of mean closing price vs month - WORKING NOW
    # new_temp = data_frame.groupby('month')['Close'].mean().plot.bar()
    # plt.title('Average Closing Price By Month')
    # plt.xlabel('Month')
    # plt.ylabel('Stock Price')
    # plt.show()
    
    close = generateCloseModel(data_frame, ticker)
    holt = HoltExp(data_frame, ticker)
    goldenCross = goldenDeathCross(data_frame, ticker)
    decision = sorted([close, holt], key=lambda x: x[2])[0]
    final = []
    if decision[0] == goldenCross:
        final.append(goldenCross)
    else:
        final.append("HOLD")
    final.append(decision[1])
    final.append(decision[2])
    final.append(decision[3])
    # return(close, holt, goldenCross)
    #return sorted([close], key=lambda x: x[2])[0]
    return tuple(final)

async def get_data(ticker, api_key):
    ts = TimeSeries(key=api_key, output_format='pandas')
    data_frame, _ = await ts.get_daily(symbol=ticker, outputsize='full')
    await ts.close()
    return data_frame




# --- NEXT TASKS ---#

# TODO: Present status using InitialState
# TODO: S&P 500 metrics
